runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,365 @@
+    end,
+    on_reload = function()
+      bufstate = { row_level = {}, row_text = {}, kind_ranges = {} }
+      bufstates[bufnr] = bufstate
+      for _, client in
+        ipairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+      do
+        refresh(bufnr, client)
+      end
+    end,
+    --- When modifying the buffer, new foldinfo is scheduled to be updated,
+    --- synchronize the changed rows with their previous foldlevels as possible,
+    --- before new foldlevels is updated.
+    ---@param start_row integer
+    ---@param old_row integer
+    ---@param new_row integer
+    on_bytes = function(_, _, _, start_row, _, _, old_row, _, _, new_row, _, _)
+      if bufstate == nil then
+        return true
+      end
+      local row = new_row - old_row
+      local row_level = bufstate.row_level
+      if vim.tbl_isempty(row_level) then
+        return
+      end
+      if row > 0 then
+        vim._list_insert(bufstate.row_level, start_row, start_row + math.abs(row) - 1, { -1 })
+      elseif row < 0 then
+        vim._list_remove(bufstate.row_level, start_row, start_row + math.abs(row) - 1)
+      end
+    end,
+  })
+  api.nvim_create_autocmd('LspDetach', {
+    group = augroup_setup,
+    buffer = bufnr,
+    callback = function()
+      local clients = vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange })
+      if #clients == 0 then
+        for _, client in ipairs(clients) do
+          refresh(bufnr, client)
+        end
+      else
+        bufstate = { row_level = {}, row_text = {}, kind_ranges = {} }
+        bufstates[bufnr] = bufstate
+        if api.nvim_buf_is_loaded(bufnr) then
+          foldupdate(bufnr)
+        end
+      end
+    end,
+  })
+  api.nvim_create_autocmd('LspNotify', {
+    group = augroup_setup,
+    buffer = bufnr,
+    callback = function(args)
+      local client = assert(vim.lsp.get_client_by_id(args.data.client_id))
+      if
+        client:supports_method(ms.textDocument_foldingRange, bufnr)
+        and (
+          args.data.method == ms.textDocument_didChange
+          or args.data.method == ms.textDocument_didOpen
+        )
+      then
+        refresh(bufnr, client)
+      end
+    end,
+  })
+
+  for _, client in
+    ipairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+  do
+    refresh(bufnr, client)
+  end
+
+  return bufstate
+end
+
+-- The "garbage collector" for `bufstate` and its associated event hooks.
+-- See the comment above `augroup_setup` function for more details.
+api.nvim_create_autocmd('OptionSet', {
+  group = augroup_setup,
+  pattern = 'foldexpr',
+  callback = function()
+    if vim.v.option_old == 'v:lua.vim.lsp.foldexpr()' then
+      for bufnr in pairs(bufstates) do
+        if
+          ---@param winid integer
+          vim.iter(vim.fn.win_findbuf(bufnr)):all(function(winid)
+            return vim.wo[winid].foldexpr ~= 'v:lua.vim.lsp.foldexpr()'
```

Comment:
This won't work if #31280 gets merged.

Reply:
I have already subscribed to this PR. If it is merged, this place will become more efficient because we can check if the expr functions are equal instead of checking strings.

Reply:
The point is that there is no reliable way of checking this, nor do I think it is necessary.

Reply:
Fine, revert everything back to the stage you are happy with and then someone else can review and merge this. If this gets merged I'll take a look at resolving the memory leaks in a follow up.

Reply:
Perhaps I'm overthinking it, and I hope that's the case. Later, I will add some test cases that I'm concerned about so that the code can be changed more safely.